add: 44/WAKU2-DANDELION (#541)
vacp2p/rfc: content/docs/rfcs/44/README.md

---
slug: 44
title: 44/WAKU2-DANDELION
name: Waku v2 Dandelion
status: raw
category: Standards Track
tags: waku/anonymity
editor: Daniel Kaiser <danielkaiser@status.im>
contributors:
---

# Abstract

This document specifies a deanonymization mitigation technique,
based on Dandelion and Dandelion++,

> for Waku Relay.

Reviewer: a link to the RFC is missing.

Author: see above

> It mitigates mass deanonymization in the multi-node (botnet) attacker model,

Reviewer: I think we have an FC for the attack models, would be good to provide a link here for the `multi-node` attacker model.

Author: Typically,  there are no references in the Abstract.

This is a more general decision we should make. What do others think?

As [RFC 7322](https://datatracker.ietf.org/doc/rfc7322/) says:

```
   Similarly, the Abstract should be complete in itself.  It will appear
   in isolation in publication announcements and in the online index of
   RFCs.  Therefore, the Abstract must not contain citations.
```

(Because we use direct links, we could  put links into the abstract.)

Author: I added links in 9bc2058a5
Imo, it is OK for a raw version and it would be better for later versions to rephrase so that refs are not needed.

# Background and Motivation

[Waku Relay](/spec/11/), offers privacy, pseudonymity, and a first layer of anonymity protection by design.
Being a modular protocol family [Waku v2](/spec/10/)
offers features that inherently carry trade-offs as separate building blocks.
Anonymity protection is such a feature.
The [Anonymity Trilemma](https://freedom.cs.purdue.edu/projects/trilemma.html)
states that an anonymous communication network can only have two out of
low bandwidth consumption, low latency, and strong anonymity.
Even when choosing low bandwidth and low latency, which is the trade-off that basic Waku Relay takes,
better anonymity properties (even though not strong per definition) can be achieved by sacrificing some of the efficiency properties.
44/WAKU2-DANDELION specifies one such technique, and aims at gaining the best "bang for the buck"
in terms of efficiency paid for anonymity gain.
44/WAKU2-DANDELION is based on [Dandelion](https://arxiv.org/abs/1701.04439)
and [Dandelion++](https://arxiv.org/abs/1805.11060).

Dandelion is a message spreading method, which, compared to other methods,
increases the uncertainty of an attacker when trying to link messages to senders.

> Libp2p gossipsub aims at spanning a d-regular graph topology, with d=6 as the [default value](/spec/29/#gossipsub-v10-parameters).

Reviewer: A link to the `d-regular graph topology` would be useful.

Author: 3f23234bb00

Messages are forwarded within this (expected) symmetric topology,
which reduces uncertainty when trying to link messages to senders.
Dandelion breaks this symmetry by subdividing message spreading into a "stem" and a "fluff" phase.

In the "stem" phase, the message is sent to a single select relay node.
With a certain probability, this message is relayed further on the "stem",
or enters the fluff phase.
On the stem, messages are relayed to single peers, respectively,
while in fluff phase, messages are spread as per usual relay operation (augmented by random delays to further reduce symmetry).

There are several design choices to be made in the stem phase:

1) the probability of continuing the stem phase,
2) the out degree in the stem phase, which set to 1 in this document (also in the Dandelion papers),
3) the rate of re-selecting stem relays among all gossipsub mesh peers (for a given pubsub topic), and

> 4) the mapping of incoming connections to outgoing connections.

Reviewer: It is not clear how and why it is a design choice?  it sounds like some internal affairs of the node to infer something

Author: These are choices that should be fixed per network, and not for nodes to decide individually.
I changed "in the stem phase" to "for the stem phase of Dandelion based specification" to make that clearer (59547d0b941).
I agree, the term "design" choice is a bit heavy here, but
as the paper states, these choices have a big influence on the anonymity properties.
We could call them parameter choices. Point 4), however, is not really a parameter choice.
I'd not call it simply "choice" because that would indicate a choice by nodes made individually, imo.

Dandelion++ and this document, for instance, fix two stem relays.
The resulting 2-regular overlay graph is referred to as the anonymity graph.

In the "fluff" phase the relay introduces a random delay before regularly forwarding the messages to all mesh peers.
As Dandelion is designed as an update to the Bitcoin network using diffusion spreading,
which already introduces random delays,
the Dandelion paper describes the fluff phase as regular forwarding.

Note: This is an early raw version of the specification.
It does not strictly follow the formally evaluated Dandelion++ paper,
as we want to experiment with relaxing (and strengthening) certain model properties.
In specific, we aim at a version that has tighter latency bounds.

> [Further reasearch](https://arxiv.org/pdf/2201.11860.pdf) suggests that Dandelion++'s design choices are not optimal,

Reviewer 2: ```suggestion
[Further research](https://arxiv.org/pdf/2201.11860.pdf) suggests that Dandelion++'s design choices are not optimal,
```

which further assures that tweaking design choices makes sense.
We will refine design decisions in future versions of this specification.

Further information on Waku anonymity may be found in our [Waku Privacy and Anonymity Analysis](https://vac.dev/wakuv2-relay-anon).

# Theory and Functioning

> 44/WAKU2-DANDELION can be seen an anonymity enhancing add-on to [Waku Relay](https://rfc.vac.dev/spec/11/) message dissemination,

Reviewer 2: ```suggestion
44/WAKU2-DANDELION can be seen as an anonymity enhancing add-on to [Waku Relay](https://rfc.vac.dev/spec/11/) message dissemination,
```

which is based on [libp2p gossipsub](https://github.com/libp2p/specs/blob/master/pubsub/gossipsub/README.md).
44/WAKU2-DANDELION subdivides message dissemination into a "stem" and a "fluff" phase.
This specification is mainly concerned with specifying the stem phase.
The fluff phase corresponds to [Waku Relay](/spec/11/),
except that it adds a random delay before relaying a message.

*Note: In a future version of this specification,
we might move the specification of the fluff phase into a separate document.
We envision stem and fluff phase as abstract concepts.
The Dandelion stem and fluff phases instantiate these concepts.
Future stem specifications might comprise: none (standard relay), Dandelion, Tor, and mix-net.
As for future fluff specifications: none (standard relay), diffusion (random delays), and mix-net.*

Messages relayed by nodes supporting 44/WAKU2-DANDELION are either in stem phase or in fluff phase.
We refer to the former as a stem message and to the latter as a fluff message.
A message starts in stem phase, and at some point, transitions to fluff phase.
Nodes, on the other hand, are in stem state or fluff state.

> Nodes in stem state relay stem messages to a single, per epoch and incoming stem connection randomly selected, relay node.

Reviewer 2: I only understood this sentence after I read the rest of the document. Couldn't a clearer phrasing be something like:
```suggestion
Nodes in stem state relay stem messages to a single relay node, randomly selected per epoch for each incoming stem connection.
```

Author: Yes. Thank you :).

This does not include relaying messages originated in $v$, for which $v$ SHOULD always be in stem state.

## Choosing the State

On startup and when a new epoch starts,
node $v$ randomly selects a number $r$ between 0 and 1.
If $r < q$, for $q = 0.2$, the node enters fluff state, otherwise, it enters stem state.

New epochs start when `unixtime` (in seconds) $\equiv 0 \mod 600$,
corresponding to 10 minute epochs.

## Stem State

On entering stem state,
nodes supporting 44/WAKU2-DANDELION MUST randomly select two nodes for each pubsub topic from the respective gossipsub mesh node set.
These nodes are referred to as stem relays.
Stem relays MUST support [19/WAKU2-LIGHTPUSH](/spec/19/).
If a chosen peer does not support [19/WAKU2-LIGHTPUSH](/spec/19/),
the node SHOULD switch to fluff state.
(We may update this strategy in future versions of this document.)

Further, the node establishes a map that maps each incoming stem connection
to one of its stem relays chosen at random (but fixed per epoch).
Incoming stem connections are identified by the [Peer IDs](https://docs.libp2p.io/concepts/peers/#peer-id/)

> of peers the node receives [19/WAKU2-LIGHTPUSH](/spec/19/) messages from.

Reviewer 2: Would all incoming lightpush messages be relayed on a stem or only those from peers that belong to the gossipsub mesh? I assume it would be impossible to pre-emptively map such peer IDs, as new peers can constantly start using lightpush.

Author: I'd suggest not distinguishing. The main reason is keeping the protocol simpler.
Nodes using lightpush to publish their messages would not be able to tell if their messages gets relayed via Waku Relay or Dandelion stem.
Imo, this is not a problem, because lightpush nodes

* anyway do not have a guarantee that a specific relay is honest
* benefit from the Dandelion anonymity properties

As soon as this draft gets merged, Ill open a PR suggesting an update to the lightpush RFC saying that messages are either relayed by Waku relay or via Dandelion stem.

Author: It is mentioned in the secturity considerations section.
Still, I made this exlicit at this point, too: f5030d4dcd

The node itself is mapped in the same way, so that all messages originated by the node are relayed via a per-epoch-fixed Dandelion relay, too.

While in stem state, nodes MUST relay stem messages to the respective stem relay.
Received fluff messages MUST be relayed as specified in the fluff state section.

The stem protocol ([19/WAKU2-LIGHTPUSH](/spec/19/)) is independent of the fluff protocol ([Waku Relay](/spec/11/)).
While in stem state, nodes MUST NOT gossip about stem messages,

> and MUST NOT send control messages related to stem messages.

Reviewer 2: Right. Trying to think if this requires altering underlying gossipsub behaviour in any way, but nodes will only start gossiping about messages once a node publishes the message on the `pubsubTopic` and it enters the fluff phase.
(Thinking out loud :) )

Author: Yes. I wanted to make that explictit. I'll add add a sentence saying that no actions have to be taken when implementing the stem phase in lightpush.

Author: I added a parenthesis explaining this: f5030d4dcd

and [Dandelion++](https://arxiv.org/abs/1805.11060).

Dandelion is a message spreading method, which, compared to other methods,
increases the uncertainty of an attacker when trying to link messages to senders.
Libp2p gossipsub aims at spanning a d-regular graph topology, with d=6 as the [default value](/spec/29/#gossipsub-v10-parameters).
Messages are forwarded within this (expected) symmetric topology,
which reduces uncertainty when trying to link messages to senders.
Dandelion breaks this symmetry by subdividing message spreading into a "stem" and a "fluff" phase.

In the "stem" phase, the message is sent to a single select relay node.
With a certain probability, this message is relayed further on the "stem",
or enters the fluff phase.
On the stem, messages are relayed to single peers, respectively,
while in fluff phase, messages are spread as per usual relay operation (augmented by random delays to further reduce symmetry).

There are several design choices to be made in the stem phase:

1) the probability of continuing the stem phase,
2) the out degree in the stem phase, which set to 1 in this document (also in the Dandelion papers),

> 3) the rate of re-selecting stem relays among all gossipsub mesh peers (for a given pubsub topic), and

Reviewer: Can you please clarify this part?

Author: This was meant as list considerations for tuning that should be removed once the RFC moves to a more stable state.
I agree, this is confusing in the introduction.
I moved it to the "future analysis". The background and specifications sections should provide the necessary context.
(The future analysis section will be removed in future versions.)

28d162008636

4) the mapping of incoming connections to outgoing connections.

> Dandelion++ and this document, for instance, fix two stem relays.

Reviewer: `fix two stem relays.` as the out degree?

Author: see next comment

The resulting 2-regular overlay graph is referred to as the anonymity graph.

In the "fluff" phase the relay introduces a random delay before regularly forwarding the messages to all mesh peers.
As Dandelion is designed as an update to the Bitcoin network using diffusion spreading,
which already introduces random delays,
the Dandelion paper describes the fluff phase as regular forwarding.

Note: This is an early raw version of the specification.
It does not strictly follow the formally evaluated Dandelion++ paper,
as we want to experiment with relaxing (and strengthening) certain model properties.
In specific, we aim at a version that has tighter latency bounds.
[Further research](https://arxiv.org/pdf/2201.11860.pdf) suggests that Dandelion++'s design choices are not optimal,
which further assures that tweaking design choices makes sense.
We will refine design decisions in future versions of this specification.

Further information on Waku anonymity may be found in our [Waku Privacy and Anonymity Analysis](https://vac.dev/wakuv2-relay-anon).

# Theory and Functioning

> 44/WAKU2-DANDELION can be seen as an anonymity enhancing add-on to [Waku Relay](https://rfc.vac.dev/spec/11/) message dissemination,

Reviewer: If later we are going to generalize the `stem` and `fluff` phases, then it may make sense to select a more general name for this spec e.g., WAKU2-NETWORK-ANONYMITY?

Reviewer: Although, personally I prefer to keep it as WAKU2-DANDELION, and have other specifications for different future anonymity solutions e.g., Tor, and Mixinet. Later we can bundle them together in a meta specification e.g., WAKU2-NETWORK-ANONYMITY.

Author: Yes. For now, I'd keep the Dandelion name.
We could name it `WAKU2-DANDELION-STEM` in the future; specific stem instantiations could be named `WAKU2-STEM-*`
Imo, the fluff specs could be integrated into Relay (and its anonymity extensions, e.g. Tor, Mix) in the future.

44/WAKU2-DANDELION subdivides message dissemination into a "stem" and a "fluff" phase.
This specification is mainly concerned with specifying the stem phase.
The fluff phase corresponds to [Waku Relay](/spec/11/),
except that it adds a random delay before relaying a message.

*Note: In a future version of this specification,
we might move the specification of the fluff phase into a separate document.
We envision stem and fluff phase as abstract concepts.
The Dandelion stem and fluff phases instantiate these concepts.
Future stem specifications might comprise: none (standard relay), Dandelion, Tor, and mix-net.
As for future fluff specifications: none (standard relay), diffusion (random delays), and mix-net.*

Messages relayed by nodes supporting 44/WAKU2-DANDELION are either in stem phase or in fluff phase.
We refer to the former as a stem message and to the latter as a fluff message.
A message starts in stem phase, and at some point, transitions to fluff phase.
Nodes, on the other hand, are in stem state or fluff state.
Nodes in stem state relay stem messages to a single relay node, randomly selected per epoch for each incoming stem connection.

> Nodes in fluff state transition stem messages into fluff phase and relay them accordingly.

Reviewer: Don't nodes in the fluff phase route fluff messages as well?

Author: Yes. This is what 

> Fluff messages are always disseminated via Waku Relay (with added random delay).

indicates.
I extended this to

> Fluff messages are always disseminated via Waku Relay (with added random delay),
by both nodes in stem state and nodes in fluff state. 

to clarify. 
79fa6cf8d5

Fluff messages are always disseminated via Waku Relay (with added random delay).

Messages originated in the node (i.e. messages coming from the application layer of our node),
are always sent as stem messages.

The stem phase can be seen as a different protocol, and messages are introduced into Waku Relay, and by extension gossipsub,
once they arrive at a node in fluff state for the first time.
44/WAKU2-DANDELION uses [19/WAKU2-LIGHTPUSH](/spec/19/) as the protocol for relaying stem messages.

There are no negative effects on gossipsub peer scoring,
because Dandelion nodes in *stem state* still normally relay Waku Relay (gossipsub) messages.

# Specification

> Nodes supporting 44/WAKU2-DANDELION MUST either be in stem state or in fluff state.

Reviewer: Nodes may have different states for different messages, right? So it is not a static state per node?

Author: The state is fixed per epoch. However, stem state nodes also relay fluff messages.
The state only effects how relay messages are relayed. 

> While in stem state, nodes MUST relay stem messages to the respective stem relay.
Received fluff messages MUST be relayed as specified in the fluff state section.

Also, adding 

> Fluff messages are always disseminated via Waku Relay (with added random delay),
by both nodes in stem state and nodes in fluff state.

above in response to your previous feedback comment should clarify this at that point.

> This does not include relaying messages originated in $v$, for which $v$ SHOULD always be in stem state.

Reviewer: What is `V`?

Author: Addressed in 5fa20b5d9

## Choosing the State

On startup and when a new epoch starts,
node $v$ randomly selects a number $r$ between 0 and 1.
If $r < q$, for $q = 0.2$, the node enters fluff state, otherwise, it enters stem state.

New epochs start when `unixtime` (in seconds) $\equiv 0 \mod 600$,
corresponding to 10 minute epochs.

## Stem State

On entering stem state,
nodes supporting 44/WAKU2-DANDELION MUST randomly select two nodes for each pubsub topic from the respective gossipsub mesh node set.
These nodes are referred to as stem relays.
Stem relays MUST support [19/WAKU2-LIGHTPUSH](/spec/19/).
If a chosen peer does not support [19/WAKU2-LIGHTPUSH](/spec/19/),
the node SHOULD switch to fluff state.
(We may update this strategy in future versions of this document.)

Further, the node establishes a map that maps each incoming stem connection
to one of its stem relays chosen at random (but fixed per epoch).
Incoming stem connections are identified by the [Peer IDs](https://docs.libp2p.io/concepts/peers/#peer-id/)
of peers the node receives [19/WAKU2-LIGHTPUSH](/spec/19/) messages from.

> Incoming [19/WAKU2-LIGHTPUSH](/spec/19/) connections from peers that do not support 44/WAKU2-DANDELION are treated in the same way.

Reviewer: Just to be consistent with the preceding sentence:
```suggestion
Incoming [19/WAKU2-LIGHTPUSH](/spec/19/) messages from peers that do not support 44/WAKU2-DANDELION are treated in the same way.
```

Author: This is referring to the mapping of connections. I clarified this by changing
"treated" to "identified and mapped"  50002213b58

This makes the protocol simpler, increases the anonymity set, and offers Dandelion anonymity properties to such peers, too.

The node itself is mapped in the same way, so that all messages originated by the node are relayed via a per-epoch-fixed Dandelion relay, too.

While in stem state, nodes MUST relay stem messages to the respective stem relay.
Received fluff messages MUST be relayed as specified in the fluff state section.

The stem protocol ([19/WAKU2-LIGHTPUSH](/spec/19/)) is independent of the fluff protocol ([Waku Relay](/spec/11/)).
While in stem state, nodes MUST NOT gossip about stem messages,
and MUST NOT send control messages related to stem messages.
(An existing gossipsub implementation does *not* have to be adjusted to not send gossip about stem messages,
 because these messages are only handed to gossipsub once they enter fluff phase.)

### Fail Safe

> Nodes $v$ in stem state SHOULD store messages attached with a random timer between $t_1 = 5 * 100ms$ and $t_2 = 2 * t_1$.

Reviewer: ```suggestion
Node $v$ in stem state SHOULD store messages attached with a random timer between $t_1 = 5 * 100ms$ and $t_2 = 2 * t_1$.
```

Author: This refers to all nodes v. (Short for: "For all nodes $v$ in the stem state set").

This time interval is chosen because

* we assume $100\,ms$ as an average per hop delay, and

> * using $q=0.2$ will lead to an expected number of 5 stem hops per message.

Reviewer: I assume `q=.2` represents the probability of some event, what is that event?

Author: It is introduced in the subsection above:  "Choosing the State":

> If $r < q$, for $q = 0.2$, the node enters fluff state, otherwise, it enters stem state.

This is a Dandelion parameter. (The paper also names it `q`).

If $v$ does not receive a given message via Waku Relay (fluff) before the respective timer runs out,
$v$ will disseminate the message via Waku Relay.

## Fluff State

In fluff state, nodes operate as usual Waku Relay nodes,

> with the exception of adding a random delay between $0$ and $100\,ms$ before relaying a message.

Reviewer: One of the scores in the peer scoring is related to the fast delivery of the messages i.e., `p2` `First Message Deliveries `, by introducing this delay, peers may lose the chance to get that score.

Author: Note added in fef435e76a3

By adding this delay, the fluff phase modifies the behaviour of [libp2p gossipsub](https://github.com/libp2p/specs/blob/master/pubsub/gossipsub/README.md),
which Waku Relay builds upon.

As mentioned in previous sections, we might move the fluff phase specification into a separate document.
We will look more into (optimal) delay ranges and might introduce dynamic delay ranges based on traffic volume.

# Implementation Notes

Handling of the 44/WAKU2-DANDELION stem phase can be implemented as an extension to an existing [19/WAKU2-LIGHTPUSH](/spec/19/) implementation.

Since the 44/WAKU2-DANDELION fluff phase alters gossipsub message dissemination,
it is implemented on the libp2p gossipsub layer.

# Security/Privacy Considerations

## Denial of Service: Black Hole Attack

In a [black hole attack](/spec/45/#black-hole-internal), malicious nodes prevent messages from being spread,

> metaphorically not allowing messages to leave once they entered.

Reviewer: just for better clarity: 
```suggestion
metaphorically not allowing messages to leave once they entered the stem phase.
```

Author: This is true for all phases, and for Relay without Dandelion, too.
Introducing a stem phase, just increases the risk.

> This requires the attacker to control nodes on all dissemination paths.

Reviewer: Does `dissemination` refer to the `stem` phase?

Author: Generally to all dissemination path. The next sentence should clarify this:

> Since the number of dissemination paths is significantly reduced in the stem phase,

Since the number of dissemination paths is significantly reduced in the stem phase,
Dandelion spreading reduces the requirements for a black hole attack.

The fail-safe mechanism specified in this document (proposed in the Dandelion paper), mitigates this.

## Anonymity Considerations

### Attacker Model and Anonymity Goals

44/WAKU2-DANDELION provides significant mitigation against mass deanonymization in the
passive [scaling multi node model](/spec/45/#scaling-multi-node).
in which the attacker controls a certain percentage of nodes in the network.
44/WAKU2-DANDELION provides significant mitigation against mass deanonymization

> even if the attacker knows the network topology, i.e. the anonymity graph and the relay mesh graph.

Reviewer: What is `the anonymity graph`?

Author: It is in the "Background section":

> The graph spanned by stem connections is referred to as the anonymity graph.

(rephrased addressing one of the feedback items above.)

> Mitigation in stronger models, including the active *scaling multi node* model, is weak.

Reviewer: ```suggestion
Mitigation in stronger models, including the *active scaling multi-node* model, is weak.
```

Stem messages relayed in stem state MUST have the stem flag set.
Fluff messages MUST be relayed as specified in the fluff state section.

### Fail Safe

Nodes $v$ in stem state SHOULD store messages attached with a random timer between $t_1 = 5 * 100ms$ and $t_2 = 2 * t_1$.
This time interval is chosen because

* we assume $100\,ms$ as an average per hop delay, and
* using $q=0.2$ will lead to an expected number of 5 stem hops per message.

If $v$ does not receive a given message via Waku relay (fluff) before the respective timer runs out,
$v$ will disseminate the message via Waku relay.

## Fluff State

In fluff state, nodes operate as usual Waku Relay nodes, with the exception of adding a random delay between $0$ and $100\,ms$ before relaying a message.
Before relaying a message in fluff state, the `stem` flag MUST be set to false.

As mentioned in previous sections, we might move the stem phase specification into a separate document.
We will look more into (optimal) delay ranges and might introduce dynamic delay ranges based on traffic volume.

# Implementation Notes

Since 44/WAKU2-DANDELION alters message dissemination,

> it is implemented on the libp2p gossipsub layer.

Reviewer 3: Maybe it would be better to do the stem phase over light push, and then fluff over GS? This way LP is also protected (I think), and no need to change the core of GS

Author: Thank you :) . Good idea!
The plan was to somehow separate the stem phase into a separate protocol.
Using light push would nicely fit that design idea. I'll look into it.

it also reduces anonymity properties.
A fixed stem length requires the message to carry information about the remaining stem length.
This information reduces the uncertainty of attackers
when calculating the probability distribution assigning each node a probability for having sent a specific message.
We will quantify the resulting loss of anonymity in future versions of this document.

### Stem Relay Selection

In its current version, 44/WAKU2-DANDELION nodes default to fluff state
if the random stem relay selection yields at least one peer that does not support [19/WAKU2-LIGHTPUSH](/spec/19/) (which is the stem protocol used in 44/WAKU2-DANDELION).
If nodes would reselect peers until they find peers supporting [19/WAKU2-LIGHTPUSH](/spec/19/),
malicious nodes would get an advantage if a significant number of honest nodes would not support [19/WAKU2-LIGHTPUSH](/spec/19/).
Even though this causes messages to enter fluff phase earlier,
we choose the trade-off in favour of protocol stability and sacrifice a bit of anonymity.
(We will look into improving this in future versions of this document.)

### Random Delay in Fluff Phase

> The random delay in the fluff phase MAY be set to 0 to achieve lower latency.

Reviewer 3: I'm a bit confused by this delay, it's said earlier that no modifications to gossipsub are required, but this delay would be a modification to gossipsub?
Also, it's not explained why it's necessary

Last point, iirc each waku message contains the originator timestamp, seems to me that this break dandelion, since I can know if you are the originator if now - timestamp < 100ms

Author: Stem phase does not alter libp2p, while fluff phase does.
The implementation section says this:

---
Since the 44/WAKU2-DANDELION fluff phase alters gossipsub message dissemination,
it is implemented on the libp2p gossipsub layer.

---

>Also, it's not explained why it's necessary

I added the following to explain why:

---
This added random delay further reduces symmetry in dissemination patterns and
indroduces more uncertainty for the attacker.

---

> Last point, iirc each waku message contains the originator timestamp, seems to me that this break dandelion, since I can know if you are the originator if now - timestamp < 100ms

Good point!
I added this to the security section. aabf9c99a478

---
This document is oblivious to the actual message content, adding anonymity on the routing layer.
Still, it is important to note that [Waku2 messages](https://rfc.vac.dev/spec/14/#payloads) (in their current version) carry an originator timestamp,
which works against fluff phase random delays.
An analysis of the benefits of this timestamp versus anonymity risks is on our roadmap.

---

Generally, stem and fluff phase can be employed independently.
This is why I suggest moving the fluff phase into a separate spec in the future.

Reviewer 2: Already merged, but fwiw I agree to leave this out of the spec for now. :)

Also, re:

>  iirc each waku message contains the originator timestamp

This timestamp is optional, precisely because of the security risks. Also, I had the idea that this timestamp should tick on something like 500ms boundaries, which will be useful enough to use it for ordering, but too coarse to be effectively used in timing attacks. I strongly suggest that we make this part of a recommendation in specs soon.
